firebase-perf/src/main/java/com/google/firebase/perf/application/FrameMetricsRecorder.java

```diff
@@ -0,0 +1,182 @@
+  }
+
+  /** Starts recording FrameMetrics for the activity window */
+  public void start() {
+    if (isRecording) {
+      logger.error(
+          "FrameMetricsAggregator is already recording %s", activity.getClass().getSimpleName());
+      return;
+    }
+    fma.add(activity);
+    isRecording = true;
+  }
+
+  /**
+   * Stops recording FrameMetrics for the activity window
+   *
+   * @return FrameMetrics accumulated during the current recording
+   */
+  public Optional<PerfFrameMetrics> stop() {
+    if (!isRecording) {
+      logger.error("Cannot stop because no recording was started");
+      return Optional.absent();
+    }
+    Optional<PerfFrameMetrics> data = this.snapshot();
+    try {
+      fma.remove(activity);
+    } catch (IllegalArgumentException err) {
+      logger.debug("View not hardware accelerated. Unable to collect FrameMetrics. %s", err);
+    }
+    fma.reset();
+    isRecording = false;
+    return data;
+  }
+
+  /**
+   * Starts a sub-trace in the current recording
+   *
+   * @param key a UI state to associate this sub-trace with. e.g.) fragment
+   */
+  public void startSubTrace(Object key) {
+    if (!isRecording) {
+      logger.warn("Cannot start sub-trace because FrameMetricsAggregator is not recording");
+      return;
+    }
+    if (subTraceMap.containsKey(key)) {
+      logger.warn(
+          "Cannot start sub-trace because one is already ongoing with the key %s",
+          key.getClass().getSimpleName());
+      return;
+    }
+    Optional<PerfFrameMetrics> snapshot = this.snapshot();
+    if (!snapshot.isAvailable()) {
+      logger.warn("startSubTrace(key): snapshot() failed");
+      return;
+    }
+    subTraceMap.put(key, snapshot.get());
+  }
+
+  /**
+   * Stops the sub-trace associated with the given UI state
+   *
+   * @param key the UI state associated with the sub-trace to be stopped
+   * @return FrameMetrics accumulated during this sub-trace
+   */
+  public Optional<PerfFrameMetrics> stopSubTrace(Object key) {
+    if (!isRecording) {
+      logger.error("Cannot stop sub-trace because FrameMetricsAggregator is not recording");
+      return Optional.absent();
+    }
+    if (!subTraceMap.containsKey(key)) {
+      logger.error(
+          "Sub-trace associated with key %s was not started or does not exist",
+          key.getClass().getSimpleName());
+      return Optional.absent();
+    }
+    PerfFrameMetrics snapshotStart = subTraceMap.remove(key);
+    Optional<PerfFrameMetrics> snapshotEnd = this.snapshot();
+    if (!snapshotEnd.isAvailable()) {
+      logger.error("stopSubTrace(key): snapshot() failed");
+      return Optional.absent();
+    }
+    return Optional.of(snapshotEnd.get().subtract(snapshotStart));
+  }
+
+  /**
+   * Calculate total frames, slow frames, and frozen frames from SparseIntArray[] recorded by {@link
+   * FrameMetricsAggregator}.
+   *
+   * @return the frame metrics
+   */
+  protected Optional<PerfFrameMetrics> snapshot() {
```

Comment:
Why is this protected instead of private?

Reply:
For testing and mocking in the tests. Ideally things like these implementation detail should be private, but I'm not sure how to test it and mock it if I do that.

Reply:
Made it private and test class behaviors instead, following go/unit-testing-practices